Resolve: refactor away Module::import_resolutions and improve import resolution

jseyfried opened the pull request:
This PR adds to `NameBinding` so it can more fully represent bindings from imports as well from items, refactors away `Target`, generalizes `ImportResolution` to a simpler type `NameResolution`, and uses a single `NameResolution`-valued map in place the existing maps `children` and `import_resolutions` (of `NameBinding`s and `ImportResolution`s, respectively), simplifying duplicate checking and name resolution.

It also unifies the `resolve_name_in_module` in `lib.rs` with its namesake in `resolve_imports.rs`, clarifying and improving the core logic (fixes #31403 and fixes #31404) while maintaining clear future-comparability with shadowable globs (i.e., never reporting that a resolution is a `Success` or is `Failing` unless this would also be knowable with shadowable globs).

Since it fixes #31403, this is technically a [breaking-change], but it is exceedingly unlikely to cause breakage in practice. The following is an example of code that would break:
```rust
mod foo {
    pub mod bar {} // This defines bar in the type namespace
    pub use alpha::bar; // This defines bar in the value namespace

    // This should define baz in both namespaces, but it only defines baz in the type namespace.
    pub use self::bar as baz;
    pub fn baz() {} // This should collide with baz, but now it does not.
}

pub fn f() {}
mod alpha {
    pub use self::f as bar; // Changing this to `pub fn bar() {}` causes the collision right now.
    pub use super::*;
}
```

r? @nrc

jseyfried commented:
cc @petrochenkov @nikomatsakis

petrochenkov commented:
>single NameResolution-valued map in place the existing maps children and import_resolutions

:tada:
>fixes #31403 and fixes #31403

The issue numbers are the same.

I'll read this today.

jseyfried commented:
> The issue numbers are the same.

The second one was supposed to be #31404, edited.
> I'll read this today.

Awesome, I look forward to any feedback.

petrochenkov commented:
Ok, I give up, too many things are happening here at the same time.
I'm happy with the direction to unifying `use` items with other items, but can't follow many concrete changes, especially that I'm not very familiar with the import resolution code.

By the way, how much of the code treating imports and extern crates specially is caused only by error message compatibility?

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -788,6 +792,59 @@ enum ParentLink<'a> {
+    // Define the name or return the existing binding if there is a collision.
+    fn try_define(&mut self, binding: &'a NameBinding<'a>) -> Result<(), &'a NameBinding<'a>> {
+        let old_binding = match self.binding {
+            Some(_) if binding.defined_with(DefModifiers::SHADOWABLE) => return Ok(()),
```
So, if we try to define a shadowable binding (1) conflicting with another existing shadowable binding (2), then (2) wins or the conflict is reported somewhere else?

jseyfried replied:
The binding (1) would win (since the first match arm would be taken) and the conflict would not be reported (this is the same behavior as in the original code). This isn't a problem since the only `SHADOWABLE` bindings are from the special prelude import, so there can't be conflicts.

jseyfried replied:
Perhaps I should rename SHADOWABLE to PRELUDE.

petrochenkov replied:
>Perhaps I should rename SHADOWABLE to PRELUDE.

That would be much clearer, because prelude and shadowable glob imports would behave differently (I suppose).
The resolution result depending on the resolution order is still pretty bad, even if the concern is theoretical at the moment due to absence of user-defined preludes.

jseyfried replied:
True, I'll make it an error to have duplicate SHADOWABLE / PRELUDE imports.

jseyfried commented:
> too many things are happening here at the same time

I'll split the first commit into more smaller commits.
> how much of the code treating imports and extern crates specially is caused only by error message compatibility?

The special treatment in `report_conflicts` is for error message compatibility and the special treatment in `record_use` is for compatibility with unused extern crate lint system (which itself needs some refactoring; it tracks extern crates themselves, and not `extern crate` items so that if the same extern crate is declarated twice in different modules and only one declaration is used, the unused declaration will not be reported).
Apart from that, the special treatment is a reflection of semantics.

jseyfried commented:
I addressed @gereeter's and @petrochenkov's comments and rewrote history to make the [biggest commit](https://github.com/jseyfried/rust/commit/b85efc4d24bba24b422b4cc53f80b693faa30d5c) a little cleaner and easier to review.

jseyfried commented:
r? @nrc

petrochenkov commented:
Reviewed everything except for https://github.com/jseyfried/rust/commit/7000e708252bd9e45f28926d8fd38cc9ec054e57

nrc commented:
@bors: r+

bors commented:
:pushpin: Commit 3df40c0 has been approved by `nrc`

<!-- @bors r=nrc 3df40c09ec5efb74e16c3da84fef6ce142dae73a -->

bors commented:
:hourglass: Testing commit 3df40c0 with merge 1de70d3...

jseyfried commented:
@nrc Thanks!

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8025), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8046), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/1293), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7986), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7996), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7986), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1199), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3151), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8031), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8034), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7999), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/169), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3039), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3055), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3046), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3028), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2261), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3033)
